ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/LazyObjectIterable.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * ******************************************************************************
+ *   Copyright 2014-2015 Spectra Logic Corporation. All Rights Reserved.
+ *   Licensed under the Apache License, Version 2.0 (the "License"). You may not use
+ *   this file except in compliance with the License. A copy of the License is located at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file.
+ *   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ *   CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *   specific language governing permissions and limitations under the License.
+ * ****************************************************************************
+ */
+
+package com.spectralogic.ds3client.helpers;
+
+import com.spectralogic.ds3client.Ds3Client;
+import com.spectralogic.ds3client.commands.GetBucketRequest;
+import com.spectralogic.ds3client.commands.GetBucketResponse;
+import com.spectralogic.ds3client.models.Contents;
+import com.spectralogic.ds3client.models.ListBucketResult;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Produces an Iterator that will list contents from the get bucket call lazily.
+ * Our definition of lazy for this class, is that the Iterator is returned right
+ * away with no data.  Data is only then loaded, as needed, in pieces.
+ */
+public class LazyObjectIterable implements Iterable<Contents> {
+
+    private static final int DEFAULT_MAX_KEYS = 1000;
+    private final Ds3Client client;
+    private final String bucket;
+    private final String prefix;
+    private final String nextMarker;
+    private final int maxKeys;
+    private final int retryCount;
+
+    public LazyObjectIterable(final Ds3Client client, final String bucket, final String keyPrefix, final String nextMarker, final int maxKeys, final int retryCount) {
```

Comment:
The reason I think it should is that the helper function is currently where all that information exists.  And rather than duplicating the telescoping constructors over into the iterator, it's all kept in one place.